Wallet - negative balance

When trying to send transaction during the synchronization process, it is possible to achieve negative balance in wallet
```
13:12:04.520 INFO  [tor.default-dispatcher-23] o.e.n.w.ErgoWalletActor - Processing block at height 4018 with 1 wallet transactions
[ERROR] [07/24/2019 13:12:04.542] [ergoref-akka.actor.default-dispatcher-24] [akka://ergoref/user/$h] requirement failed: Balance could not be negative
java.lang.IllegalArgumentException: requirement failed: Balance could not be negative
        at scala.Predef$.require(Predef.scala:281)
        at org.ergoplatform.nodeView.wallet.persistence.WalletRegistry.$anonfun$updateOnBlock$11(WalletRegistry.scala:116)
        at org.ergoplatform.nodeView.wallet.persistence.RegistryOps$.$anonfun$updateIndex$1(RegistryOps.scala:112)
        at cats.free.Free.$anonfun$step$1(Free.scala:52)
        at cats.free.Free.$anonfun$foldMap$3(Free.scala:156)
        at cats.Monad.$anonfun$map$1(Monad.scala:16)
        at cats.data.IndexedStateT.$anonfun$flatMap$3(IndexedStateT.scala:28)
        at cats.Eval$.loop$1(Eval.scala:363)
        at cats.Eval$.cats$Eval$$evaluate(Eval.scala:368)
        at cats.Eval$FlatMap.value(Eval.scala:307)
        at org.ergoplatform.nodeView.wallet.persistence.RegistryOps$TransactionOp.transact(RegistryOps.scala:36)
        at org.ergoplatform.nodeView.wallet.persistence.RegistryOps$TransactionOp.transact(RegistryOps.scala:33)
        at org.ergoplatform.nodeView.wallet.persistence.WalletRegistry.updateOnBlock(WalletRegistry.scala:121)
        at org.ergoplatform.nodeView.wallet.ErgoWalletActor.org$ergoplatform$nodeView$wallet$ErgoWalletActor$$processBlock(ErgoWalletActor.scala:421)
        at org.ergoplatform.nodeView.wallet.ErgoWalletActor$$anonfun$scanLogic$1.applyOrElse(ErgoWalletActor.scala:123)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:176)
        at akka.actor.Actor.aroundReceive(Actor.scala:539)
        at akka.actor.Actor.aroundReceive$(Actor.scala:537)
        at org.ergoplatform.nodeView.wallet.ErgoWalletActor.aroundReceive(ErgoWalletActor.scala:35)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:612)
        at akka.actor.ActorCell.invoke(ActorCell.scala:581)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
        at akka.dispatch.Mailbox.run(Mailbox.scala:229)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
After this requirement failure wallet synchronization stale at this height and never updates

--- a/src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala
+++ b/src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala
@@ -83,15 +83,13 @@ final class WalletRegistry(store: Store)(ws: WalletSettings) extends ScorexLoggi
       _ <- putBoxes(certainBxs ++ uncertainBxs)
       _ <- putTxs(txs)
       spentBoxesWithTx <- getBoxes(inputs.map(x => decodedBoxId(x._2))).map(
-        _.flatten.filter(_.certainty.certain).flatMap(bx =>
-          inputs.find(_._2 == encodedBoxId(bx.box.id)).map { case (txId, _) => txId -> bx })
+        _.flatten.flatMap(bx => inputs.find(_._2 == encodedBoxId(bx.box.id)).map { case (txId, _) => txId -> bx })
       )
       _ <- processHistoricalBoxes(spentBoxesWithTx, blockHeight)
       _ <- updateIndex { case RegistryIndex(_, balance, tokensBalance, _) =>
-        log.info(s"SpentBoxes at (h: $blockHeight) *> $spentBoxesWithTx")
-        val spentBoxes = spentBoxesWithTx.map(_._2)
-        val spentAmt = spentBoxes.map(_.box.value).sum
-        val spentTokensAmt = spentBoxes
+        val spentCertainBoxes = spentBoxesWithTx.map(_._2).filter(_.certainty.certain)
+        val spentAmt = spentCertainBoxes.map(_.box.value).sum
+        val spentTokensAmt = spentCertainBoxes
           .flatMap(_.box.additionalTokens.toArray)
           .foldLeft(Map.empty[EncodedTokenId, Long]) { case (acc, (id, amt)) =>
             acc.updated(encodedTokenId(id), acc.getOrElse(encodedTokenId(id), 0L) + amt)

--- a/src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala
+++ b/src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala
@@ -82,11 +82,13 @@ final class WalletRegistry(store: Store)(ws: WalletSettings) extends ScorexLoggi
     val update = for {
       _ <- putBoxes(certainBxs ++ uncertainBxs)
       _ <- putTxs(txs)
-      spentBoxesWithTx <- getBoxes(inputs.map(x => decodedBoxId(x._2))).map(_.flatten.flatMap(bx =>
-        inputs.find(_._2 == encodedBoxId(bx.box.id)).map { case (txId, _) => txId -> bx })
+      spentBoxesWithTx <- getBoxes(inputs.map(x => decodedBoxId(x._2))).map(
+        _.flatten.filter(_.certainty.certain).flatMap(bx =>
+          inputs.find(_._2 == encodedBoxId(bx.box.id)).map { case (txId, _) => txId -> bx })
       )
       _ <- processHistoricalBoxes(spentBoxesWithTx, blockHeight)
       _ <- updateIndex { case RegistryIndex(_, balance, tokensBalance, _) =>
+        log.info(s"SpentBoxes at (h: $blockHeight) *> $spentBoxesWithTx")
         val spentBoxes = spentBoxesWithTx.map(_._2)
         val spentAmt = spentBoxes.map(_.box.value).sum
         val spentTokensAmt = spentBoxes

--- a/src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala
+++ b/src/test/scala/org/ergoplatform/nodeView/wallet/ErgoWalletSpec.scala
@@ -1,22 +1,24 @@
 package org.ergoplatform.nodeView.wallet
 
 import org.ergoplatform._
-import org.ergoplatform.modifiers.mempool.{ErgoBoxSerializer, ErgoTransaction}
+import org.ergoplatform.modifiers.mempool.{ErgoBoxSerializer, ErgoTransaction, UnsignedErgoTransaction}
 import org.ergoplatform.nodeView.state.{ErgoStateContext, VotingData}
 import org.ergoplatform.nodeView.wallet.IdUtils._
 import org.ergoplatform.nodeView.wallet.persistence.RegistryIndex
 import org.ergoplatform.nodeView.wallet.requests.{AssetIssueRequest, PaymentRequest}
 import org.ergoplatform.settings.{Constants, LaunchParameters}
 import org.ergoplatform.utils._
-import org.ergoplatform.wallet.interpreter.ErgoInterpreter
+import org.ergoplatform.wallet.interpreter.{ErgoInterpreter, ErgoUnsafeProver}
 import org.scalatest.PropSpec
 import scorex.crypto.authds.ADKey
 import scorex.crypto.hash.Blake2b256
 import scorex.util.encode.Base16
 import sigmastate.Values.ByteArrayConstant
 import sigmastate._
+import sigmastate.basics.DLogProtocol.DLogProverInput
 import sigmastate.eval._
 import sigmastate.eval.Extensions._
+import sigmastate.Values._
 
 import scala.concurrent.blocking
 import scala.util.Random
@@ -26,6 +28,53 @@ class ErgoWalletSpec extends PropSpec with WalletTestOps {
   private implicit val verifier: ErgoInterpreter = ErgoInterpreter(LaunchParameters)
   private implicit val ergoAddressEncoder: ErgoAddressEncoder =
     new ErgoAddressEncoder(settings.chainSettings.addressPrefix)
+
+  property("uncertain boxes spending") {
+    withFixture { implicit w =>
+      val walletPk = getPublicKeys.head.pubkey
+
+      val foreignSecret = DLogProverInput.random()
+      val foreignPk = foreignSecret.publicImage
+      val uncertainProp = SigmaOr(SigmaAnd(walletPk, GE(IntConstant(0), IntConstant(10))), foreignPk)
+
+      val genesisBlock = makeGenesisBlock(walletPk, randomNewAsset)
+
+      applyBlock(genesisBlock) shouldBe 'success
+      waitForScanning(genesisBlock)
+
+      val confirmedBalance = getConfirmedBalances.balance
+      val balanceToMakeUncertain = confirmedBalance - 1000000
+      val balanceAfterSpending = confirmedBalance - balanceToMakeUncertain
+
+      val req = PaymentRequest(Pay2SAddress(uncertainProp), balanceToMakeUncertain, Seq.empty, Map.empty)
+      val tx = await(wallet.generateTransaction(Seq(req))).get
+
+      val block = makeNextBlock(getUtxoState, Seq(tx))
+      applyBlock(block) shouldBe 'success
+      waitForScanning(block)
+
+      val index = await(wallet.confirmedBalances)
+
+      index.uncertainBoxes.size shouldBe 1
+      index.balance shouldBe balanceAfterSpending
+
+      val uncertainTxUnsigned = new UnsignedErgoTransaction(
+        index.uncertainBoxes.map(id => new UnsignedInput(decodedBoxId(id))).toIndexedSeq,
+        IndexedSeq(),
+        IndexedSeq(new ErgoBoxCandidate(balanceToMakeUncertain, Constants.TrueLeaf, block.height + 1))
+      )
+      val uncertainTx = ErgoUnsafeProver.prove(uncertainTxUnsigned, foreignSecret)
+
+      val finalBlock = makeNextBlock(getUtxoState, Seq(ErgoTransaction(uncertainTx)))
+      applyBlock(finalBlock) shouldBe 'success
+      waitForScanning(finalBlock)
+
+      val finalIndex = await(wallet.confirmedBalances)
+
+      finalIndex.uncertainBoxes shouldBe empty
+      finalIndex.balance shouldBe balanceAfterSpending
+    }
+  }
 
   property("do not use inputs spent in off-chain transaction") {
     withFixture { implicit w =>
@@ -41,7 +90,7 @@ class ErgoWalletSpec extends PropSpec with WalletTestOps {
       // prepare a lot of inputs
       val inputsToCreate = 50
       val sumToSpend = snap.balance / (inputsToCreate + 1)
-      val req = (0 until inputsToCreate).map(a => PaymentRequest(addresses.head, sumToSpend, Seq.empty, Map.empty))
+      val req = (0 until inputsToCreate).map(_ => PaymentRequest(addresses.head, sumToSpend, Seq.empty, Map.empty))
       log.info(s"Confirmed balance $snap")
       log.info(s"Payment request $req")
       val tx = await(wallet.generateTransaction(req)).get
@@ -99,7 +148,6 @@ class ErgoWalletSpec extends PropSpec with WalletTestOps {
     withFixture { implicit w =>
       val pubKey = getPublicKeys.head.pubkey
       val genesisBlock = makeGenesisBlock(pubKey, randomNewAsset)
-      val genesisTx = genesisBlock.transactions.head
       val initialBoxes = boxesAvailable(genesisBlock, pubKey)
 
       val boxesToUseEncoded = initialBoxes.map { box =>

